src/compiler/scala/tools/nsc/typechecker/Contexts.scala

```diff
@@ -1194,7 +1194,8 @@ trait Contexts { self: Analyzer =>
     }
 
     final def lookupCompanionOf(original: Symbol): Symbol = {
```

Comment:
My head started spinning with indirection, so I figure I'll rubber-duck-review this by doing some inlining...

I tried to push all logic into `lookupCompanionOf`:

```
    final def lookupCompanionOf(original: Symbol): Symbol =
      if (!original.exists) NoSymbol else {
        original.failIfStub()

        if (original.needsFlatClasses) original.info // use initialize() ?
        val flatOwnerInfo = original.owner.rawInfo

        // can we find it using the owner structure -- i.e., is it a member of some other symbol?
        val member =
          original match {
            case mcs: ModuleClassSymbol if mcs.module ne null => mcs.module
            case _ =>
              flatOwnerInfo.decl(original.name.companionName).suchThat(sym =>
                sym != original && (if (original.isClass) sym.isModule else sym.isClass) && sym.isCoDefinedWith(original)
              )
          }

        // can we decide which branch to take here directly rather than by trying the search?
        // i.e., is it a term-owned symbol??
        if (member == NoSymbol) 
          lookupScopeEntry(original) match {
            case null => NoSymbol
            case entry => entry.owner.lookupCompanion(original)
          }
        else member
      }

```